### ISSUE TITLE
[Side Project] Gamemode Achievements

### ISSUE THREAD

**Issue by func-octo** [Zet0rz/nzombies#207]
I realize that checking for achievements can be difficult to make and take resources to perform, but to keep players playing, games have achievements. Take a look at Hide and Seek for example. It has achievements inside of it's gamemode. It would be cool to have achievements (even if they were basic) in nZombies.

EDIT: If you decide this would be a good idea, but there isn't much you'd want to do with it, etc., I could start doing it if you'd like, but I could only doo basic ones like milestones for kills, perks gotten, points wasted, door bought, etc..

**Comment by Zet0rz** [comment]
It could actually be interesting. Especially because I want to make the main story's final part to require the other ones. Using a way to check for achievements would be cool, then progress could be saved by giving an achievement for completing it. It had to be saved in a .txt anyway.

We'll just have to remember that if they are to be saved on the client rather than a particular server, then clients can always edit their "achievement" file to give themselves them. In my opinion that doesn't matter though, if players really want to skip to the last easter egg, then let them.

**Comment by minin43** [comment]
I'll most likely be hosting a server that runs this gamemode once it's released. We could definitely implement this on our server.

**Comment by func-octo** [comment]
> We'll just have to remember that if they are to be saved on the client rather than a particular server, then clients can always edit their "achievement" file to give themselves them. In my opinion that doesn't matter though, if players really want to skip to the last easter egg, then let them.

I agree. This is also good because if an achievement breaks, they can still play it because they can give it o themselves until we fix it.

**Comment by Zet0rz** [comment]
> I'll most likely be hosting a server that runs this gamemode once it's released. We could definitely implement this on our server.

The way I'm thinking is clientside saved achievements, so they don't reset on you just for joining another server (or hosting another single player listen one). Achievements serverside would have to follow a completely different system using the database rather than local .txt files. We could do that too, but it'd be a completely seperate thing that otherwise wasn't needed.

**Comment by func-octo** [comment]
I just thought, what if we could implement pointshop 2 with these achievements? We could let the server owners set up a lua/txt file with an achievement name, and if they have that achievement, they could get an 'exclusive' item from the pointshop :)

**Comment by func-octo** [comment]
What are the official maps for the 'story'? I just wanted to know so I could come up with cool names for completion achievements. Also these achievements require less work because an image is not required, but maybe you could have a gui in the `nzsettings` menu to show all achievements :)

**Comment by func-octo** [comment]
Here is a possible blueprint (300x400):
![achievements1](https://cloud.githubusercontent.com/assets/10675391/15488377/b78b2dcc-2123-11e6-8526-c11a1b1fd9a7.png)

The checkmark could be a preloaded png (as above) that can be placed at the same x and y (fixed position) as the image. On top of it all, this system could be modular/object oriented.

**Comment by minin43** [comment]
> Saved on the client

Why not both?

**Comment by Zet0rz** [comment]
> Why not both?

Because

> Achievements serverside would have to follow a completely different system [...] We could do that too, but it'd be a completely seperate thing that otherwise wasn't needed.

The only reason this could work is because we needed to save data on the client anyway; then we might as well make it an achievement system. We could, as said, do both, but then it'd be seperate and most likely an addon that will appear post-release.

**Comment by func-octo** [comment]
@Zet0rz Could I work on images for the achievements (if there are any) as well as naming them?

**Comment by func-octo** [comment]
Like for Crisis Completion it could be `Pastlife Crisis` or `Unlife Crisis`
Rooftops could be `Tightrope Trouble` (Lol couldn't come up with an alliteration)

**Comment by Zet0rz** [comment]
I'm likely gonna name the achievements the same as the name of the easter egg. Which for Kosovos is planned to be "Security Protocol: 0" and for Crisis it'd likely be "The One in Control".

**Comment by func-octo** [comment]
Still alive?

**Comment by Zet0rz** [comment]
The gamemode development was put on hold while I was competing in the 2016 Gamemode Competition. That is over now but I am currently taking a break while also working on projects in uni. Will resume development after that break, probably some time late this week or early next week.

**Comment by func-octo** [comment]
Ok :) No need to rush to get back to developing it, since we have a working version 👍

**Comment by minin43** [comment]
Were you gonna make this client-side or optionally server-side? I can probably compile a list of generic achievements.

**Comment by Zet0rz** [comment]
Not sure if achievements is going to be a thing at all, but if it was, it'd be mixed - the server would handle giving achievements (along with ability to add their own), and clients would probably save them. Clients could just cheat them, but that wouldn't really do anything.

**Comment by minin43** [comment]
Idea: MP achievements are saved on a server-by-server basis, you have an achievement level viewable when  pressing tab, and you can look at other player's completed achievements along with your own. To me, what's the point of getting the achievements if you can't show them off to other people? You can't save them locally and *not* expect people to cheat them, and then it ruins the point of the achievements.
